emails: Add third welcome email based on organization's type.

laurynmm opened the pull request:
Updated version of #24626, which adds a third welcome email for new users. The email is based on the organization's type and links to one of the corporate `/for/...` pages as a guide to using Zulip. If the organization type does not have a relevant guide, then no email is scheduled or sent.

**Notes**:
- **Prep commit**: There is a prep commit that updates the logic for calculating the number of days a scheduled email is delayed so that it is not sent on a weekend (Saturday or Sunday). Previously, the logic was hard coded for a 2 day delay. Now, the function can be used to delay for any number of days.
- **Data structure for variable strings/links**: Because these emails are not sent for all organization types, instead of adding the strings / URLs to the `Realm.ORG_TYPES` data structure in `models.py`, a new data structure is created in `email_notifications.py` with these values.
  - It would be easy to update these changes to add these fields to `Realm.ORG_TYPES` instead. A question about that implementation would be whether the organization types that don't receive these emails should have empty strings for those keys or None.
- **Translation**: For translation purposes, the full string for the email subject and button text are in the data structure. Even though the strings are similar/repetitive, we need to translate the full string at once in the email and not concatenate the organization type into the partial string as a variable in the email templates.
  - There are screenshots below with the business organization type strings replaced with a string that has already been translated into the project (in Russian and German) to confirm that the strings being update for each organization type will be translated correctly. The untranslated strings in those screenshots are normal strings in the template and should not be an issue when translating.
- **Text wrapping on buttons**: The screenshots below are all full screen width. I didn't test for narrow screens for the button wrapping of the text. I imagine if there are issues with how the text wraps, then they would be present for any button text that is longer than two or three words.
- **Email name**: The current name for the new email is `followup_zulip_guide`. This follows the format of the other welcome emails, `followup_day1` and `followup_day2`.
  - Alternatively, we could use `followup_day4`.
  - Or we could update the existing welcome emails to have more descriptive names, e.g. `followup_account_info` for day 1 and `followup_zulip_topics` for day 2.
- **Guides**: Noting here that the [communities `/for/` guide](https://zulip.com/for/communities/) is pretty different from the other ones that are linked to.

---

**Screenshots and screen captures:**

- Each has the HTML and plain text version from the dev environment.
- You can see the link text in the plain text version.
- The translated screenshots subbed an existing translated string in the two parts of the template that are updated based on the organization type: email subject and email button text.

<details>
<summary>Business org</summary>

![Screenshot from 2023-03-17 12-07-07](https://user-images.githubusercontent.com/63245456/225916466-37f508fd-9979-462a-838f-384169914536.png)
![Screenshot from 2023-03-17 12-06-52](https://user-images.githubusercontent.com/63245456/225916456-14eb0f74-03ee-4caf-a151-dfce075d4322.png)
</details>
<details>
<summary>Open-source org</summary>

![Screenshot from 2023-03-17 12-08-04](https://user-images.githubusercontent.com/63245456/225916532-51b376c7-d98e-4e35-a7c6-5b5edb065830.png)
![Screenshot from 2023-03-17 12-08-16](https://user-images.githubusercontent.com/63245456/225916552-e615ba9f-c92c-4100-8b63-7155f8e66ee2.png)
</details>
<details>
<summary>Education org</summary>

![Screenshot from 2023-03-17 12-12-45](https://user-images.githubusercontent.com/63245456/225916613-977b021b-3428-4e7e-bd96-b86945f10734.png)
![Screenshot from 2023-03-17 12-12-54](https://user-images.githubusercontent.com/63245456/225916620-ca095bb9-2402-4a9b-ae81-f4675cafaa79.png)
</details>
<details>
<summary>Research org</summary>

![Screenshot from 2023-03-17 12-13-42](https://user-images.githubusercontent.com/63245456/225916663-b9fb73fa-7a0d-49ec-9a84-39a51c05b76e.png)
![Screenshot from 2023-03-17 12-13-52](https://user-images.githubusercontent.com/63245456/225916700-96691045-d2a6-46ca-92e8-3696566e8d93.png)
</details>
<details>
<summary>Event org</summary>

![Screenshot from 2023-03-17 12-14-31](https://user-images.githubusercontent.com/63245456/225916776-ee2f2633-d92c-41e9-ac9c-7306fae4b4fd.png)
![Screenshot from 2023-03-17 12-14-38](https://user-images.githubusercontent.com/63245456/225916780-e7a9f904-1975-4d90-9b4e-140986bf9db5.png)
</details>
<details>
<summary>Non-profit org</summary>

![Screenshot from 2023-03-17 12-15-26](https://user-images.githubusercontent.com/63245456/225916843-58328774-c836-4669-81d7-b7f8f1f6e042.png)
![Screenshot from 2023-03-17 12-15-34](https://user-images.githubusercontent.com/63245456/225916855-1ec75769-675a-4819-88be-a80c90572ada.png)
</details>
<details>
<summary>Community org</summary>

![Screenshot from 2023-03-17 12-16-25](https://user-images.githubusercontent.com/63245456/225916894-deffd706-4475-424c-ada6-6762a53d9f81.png)
![Screenshot from 2023-03-17 12-16-33](https://user-images.githubusercontent.com/63245456/225916925-af5403f4-78e5-4572-ba13-38e8a340936a.png)
</details>
<details>
<summary>Translated to Russian</summary>

![Screenshot from 2023-03-17 12-24-41](https://user-images.githubusercontent.com/63245456/225916959-fee1f319-9303-4c05-9bd0-bb6df1406117.png)
![Screenshot from 2023-03-17 12-24-59](https://user-images.githubusercontent.com/63245456/225916969-a828c24b-dda4-4a08-9a18-86b1e8bdfefa.png)
</details>
<details>
<summary>Translated to German</summary>

![Screenshot from 2023-03-17 12-26-52](https://user-images.githubusercontent.com/63245456/225917002-727179f1-e1f3-4a50-9bf8-a01b89c54ed4.png)
![Screenshot from 2023-03-17 12-26-59](https://user-images.githubusercontent.com/63245456/225917013-b91b0a0f-a72c-4e93-bf13-8ffd8a206dea.png)
</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
As you noted, the `/communities` page doesn't work great for non-admin users, so let's limit the emails to point to it to be admin-only for now. We should make a note to remove this restriction when that page is updated, but I don't think we'll be working on that right away.

alya commented:
> Email name

I think it would be good to move away from `_dayX` naming for these emails. We should make it easy to change the timing of these emails without having to rename files. (The proposed name seems fine to me.)

This can be separate from this PR, but we could rename `followup_day2` to something like `followup_topics` as a follow-up (hah) to #24646.

I think `followup_day1` shouldn't really be called a `followup` at all, since it's a core part of the registration flow (and for that reason doesn't include the `[Manage email preferences](http://localhost:9991/#settings/notifications) | [Unsubscribe from welcome emails](http://localhost:9991/accounts/unsubscribe/welcome/3afre4hfdxqqqgyesdjve3po)` footer).

alya commented:
Let's change the button text `Zulip's guide` -> `Zulip guide`, matching the subject.

alya commented:
I haven't manually tested, but looks good to me otherwise!

laurynmm commented:
**Updates**:
- The button text now starts with `View Zulip guide ...` to match the email subject.
  - I'll note that if we didn't have "View" in the button text, then the two strings would be the same. And that would halve the number of new strings to be translated.
- Organization types that would send emails linking to `/for/communities` are now further limited to only send to admins.
  - Added a to-do note for removing the condition. It would probably be good to have a GitHub issue in that comment if possible.
- For text wrapping in the buttons, I tested on the two longer strings: "events and conferences" and "open-source organizations". They both seem fine, so the others should be as well. See screenshots below.

**Question/clarification**:
- I'm not sure why it makes sense for the initial followup_day1 not to have those links? That email is sent no matter what and would be the first opportunity for a person to opt out of any following emails (the topics day2 one and this one).


**Screenshots on narrow screens in dev environment**:
<details>
<summary>Events email</summary>

![Screenshot from 2023-03-22 17-17-48](https://user-images.githubusercontent.com/63245456/227000156-fb6f831d-6bdd-442e-9a7a-9aaf153c98e7.png)

</details>
<details>
<summary>Open-source project email</summary>

![Screenshot from 2023-03-22 17-17-15](https://user-images.githubusercontent.com/63245456/227000150-e9501077-aeb0-45a1-8344-430f1e49ab12.png)
</details>

alya commented:
> I'm not sure why it makes sense for the initial followup_day1 not to have those links? That email is sent no matter what and would be the first opportunity for a person to opt out of any following emails (the topics day2 one and this one).

Because it's a transactional email, not a marketing email.

alya commented:
@timabbott I think this is ready for your review (you can tweak the comment to link to the issue I just filed, if desired).

alya commented:
I did some manual testing, but haven't attempted to exhaustively check everything that way.

timabbott commented on zerver/lib/email_notifications.py:
```diff
@@ -679,18 +679,19 @@ def handle_missedmessage_emails(
-def followup_day2_email_delay(user: UserProfile) -> timedelta:
-    days_to_delay = 2
+def followup_email_delay(user: UserProfile, *, days: int) -> timedelta:
+    days_to_delay = days
```
We probably should include an assert that `days >= 2` -- at least with the current algorithm, the delay function will potentially generate a delay of 0 with certain days of the week.

The other thing we'll want to be careful with in this API is that you can't do 

first_email_date = followup_email_delay(user, 2)
second_email_date = followup_email_delay(user, 3)

in that you may have both of those emails go out on a Friday.

Idunno, maybe the assert is the wrong model and we do a larger refactor here.

alya replied:
Yeah, maybe CZO is a better place to figure this out. Omar may have feedback as well.

laurynmm replied:
Updated for the refactor to schedule these onboarding emails as a set so that we can take in to account the whole flow of when theses emails are sent. This eliminates the day 1 issue for now, but allows us to adjust for that if we did ever want to add a day after registering email.

timabbott commented on zerver/lib/email_notifications.py:
```diff
@@ -700,6 +702,50 @@ def followup_email_delay(user: UserProfile, *, days: int) -> timedelta:
+        "zulip_guide_button_text": gettext_lazy("View Zulip guide for communities"),
+        "for_zulip_link": "https://zulip.com/for/communities",
+    },
+}
```
Maybe add a comment here just documenting the expected behavior if something doesn't appear here, and/or how to decide whether to add something to this list?

laurynmm replied:
Updated version does not have this dict structure any more. Instead, the `/for/.../` page url is stored in the `Realm.ORG_TYPES` dict in `models.py`.

timabbott commented on zerver/lib/email_notifications.py:
```diff
@@ -766,6 +835,17 @@ def enqueue_welcome_emails(user: UserProfile, realm_creation: bool = False) -> N
             delay=followup_email_delay(user, days=2),
         )
```
@alya just a note while reading this code path that we only send the day-2 follow up email about topics if this is your first account with a given email address. I think it's reasonable to send this one even if we aren't doing that, but wanted to flag for you.

alya replied:
Yes, I explicitly noted this decision in #24626.

timabbott commented on zerver/lib/email_notifications.py:
```diff
@@ -766,6 +835,17 @@ def enqueue_welcome_emails(user: UserProfile, realm_creation: bool = False) -> N
 
+    if send_zulip_guide_followup:
+        send_future_email(
+            "zerver/emails/followup_zulip_guide",
```
Thinking about naming, maybe we should call these `welcome_zulip_guide`? We call this category of emails "welcome emails" in a few places, and that feels more appropriate than "followup" as a thing to add more references to in the code.

I think the other sensible option is to call these `onboarding_zulip_guide`, and plan to rename things towards that. (The only thing that's nontrivial to change is the `realm_enable_welcome_emails` API field / database model name, and we could just lazily not rename that). I think it'd be a name that more obviously conveys what these are, so maybe that's the best plan? 

In other words, I propose we make this PR use `onboarding_` rather than `followup_` for all the new stuff, and we maybe do a follow-up PR to sweep all the easy things we can rename towards that, but try to not spend a lot of time on it.

laurynmm replied:
I like `onboarding_` and that plan in general.

laurynmm replied:
Updated for `onboarding_zulip_guide` for the email name.

timabbott commented on zerver/lib/email_notifications.py:
```diff
@@ -733,6 +779,29 @@ def enqueue_welcome_emails(user: UserProfile, realm_creation: bool = False) -> N
+
+    if send_zulip_guide_followup:
+        context.update(
+            zulip_guide_subject=str(
```
Hmm, I wouldn't have expected that doing the `str()` part here would be correct; I would expect that we want to let the email sending code path resolve these lazy strings from the context into strings inside `build_email`? 

Or maybe have this block be inside a `with override_language(user.default_language)` type of block.

laurynmm replied:
Updated changes so that the strings are not sent / stored as context since that cannot be translated in our current system.

Instead the strings are all marked for translation in the `onboarding_zulip_guide` email templates and a string/enum context value is passed so that only the string for the organization's type is rendered and translated in the email.

laurynmm commented:
**Updates**: _(in addition to those in the comment above)_
- Went ahead and did the refactor for the delay of these onboarding emails so that they are scheduled as a set.
  - Did the simplest option for now and just avoided sending either email on the weekend.
  - Because they are two days apart, they never are sent on the same day.
  - This eliminates the 1 day delay question / issue Tim noted in his review, and if we wanted to add an email with that particular delay, then we could be explicit on how to handle the adjustment/scheduling of all the onboarding emails.

alya commented:
That all sounds good to me!

timabbott commented on zerver/lib/email_notifications.py:
```diff
@@ -686,26 +687,58 @@ def get_onboarding_email_schedule(user: UserProfile) -> Dict[str, timedelta]:
+
+    # We don't expect to get here as realm.org_type should match an id in Realm.ORG_TYPES,
+    # but the fall through case would be to not send the onboarding_zulip_guide email.
+    return (None, "")  # nocoverage
```
We probably should have an `assert` or log an exception or something like that.

laurynmm replied:
Added a `logging.error`, but left the `# nocoverage` comment.

laurynmm replied:
Added test coverage as well.

timabbott commented on zerver/lib/email_notifications.py:
```diff
@@ -739,6 +772,25 @@ def enqueue_welcome_emails(user: UserProfile, realm_creation: bool = False) -> N
+        context.update(
+            organization_type=organization_type_reference,
+            zulip_guide_link=onboarding_zulip_guide_url,
+        )
```
I think we should probably be building a separate `context` object for this second email; it seems like a recipe for some form of sneaky bug to share a Python object between the two emails here. Maybe just a `onboarding_zulip_guide_context` variable would be pretty reasonable given that we basically always reference it on a line by itself.

One question this presents is whether we want each email to have its own unsubscribe link? I think the answer is that since they don't expire, there's probably no harm in the links being identical, but it at least deserves a comment.

laurynmm replied:
I'll split out the context for this email. Sharing the unsubscribe link makes sense to me since it doesn't expire.

Do we want to clarify the text there at all, "Unsubscribe from welcome emails"? I'm thinking that we might want to add the realm name since in theory if I signed up for another Zulip Cloud organization, I wouldn't get the topics email, but I would possibly get this Zulip guide email.

Also, since this will be the third email, I'm assuming we want to separate the context objects for the `followup_day1` and `followup_day2` as well? I can do this in the revisions to the follow_up2 email.

timabbott replied:
Let's discuss that string in chat.zulip.org; I think we might want to think about it more broadly.

And yeah, the day1/day2 emails also seem like they should have independent contexts, though we don't have to fix that to integrate this, as it's not a regression.

laurynmm commented:
**Updates**:
- Fixed the typo noted above
- Added a `logging.error` for the case where the `realm.org_type` doesn't match any id in `Realm.ORG_TYPES`
- Split out a new `context` object for this email. See [this comment](https://github.com/zulip/zulip/pull/24739#discussion_r1152070612) with two small follow-up questions.

timabbott commented:
@laurynmm I'm not sure why this test would be failing with these changes; feel free to post in #backend if you need help debugging. (I'm assuming you've just not gotten back to this PR).

laurynmm commented:
**Updates**:
- Removed the **manage email preferences** link and updated the **unsubscribe from welcome emails** link text to include the realm name. This way it's a bit more clear that the welcome emails are only unsubscribed for the specific realm. See screenshots below.
  - Sending/receiving welcome emails is not controlled by any user setting.
- Updated the welcome email flow to have two general rules: no emails on the weekend and all emails at least 2 weekdays apart. Therefore, if an email is sent on Friday, then another won't be sent until Tuesday.

**Example screenshot of email in dev environment**:

<details>
<summary>HTML version</summary>

![Screenshot from 2023-04-10 13-40-52](https://user-images.githubusercontent.com/63245456/230895079-f5e99610-7e4c-4c63-bf2c-1de6669792c3.png)
</details>
<details>
<summary>text version</summary>

![Screenshot from 2023-04-10 13-40-43](https://user-images.githubusercontent.com/63245456/230895090-812756fe-98a6-4640-83ae-46aa8415b83a.png)
</details>

timabbott commented:
This is great, merged, thanks @laurynmm!
